### PR TITLE
Board object

### PR DESCRIPTION
**Pull request by knestler**
Board object is complete and ready for review!

### CODE REVIEW

**Comment by gjcarew** [comment]
Looks good! We will have to refactor how we generate and render the board for iteration 4 but this should work just fine through iteration 3.

File: lib/board.rb [hunk 1]
```diff
@@ -6,9 +6,82 @@ class Board
...
+      "C4" => Cell.new("C4"),
+      "D1" => Cell.new("D1"),
+      "D2" => Cell.new("D2"),
+      "D3" => Cell.new("D3"),
+      "D4" => Cell.new("D4")
+     }
   end
 
+  def valid_coordinate?(cell)
+    @cells.include?(cell)
```
Comment (jasonnoble) [inline lib/board.rb:30]:
```
def valid_coordinate?(coordinate)
  cells.keys.include?(coordinate)
end
```

@cells is a hash, so we need to make sure that the coordinates is one of the cell keys.

File: lib/board.rb [hunk 2]
```diff
@@ -6,9 +6,82 @@ class Board
...
+  def stacked_ship?(placement_array)
+    empty_array = []
+    placement_array.each do |coord|
+      empty_array << @cells[coord].empty?
+    end
+    empty_array.all?(true)
+  end 
+
+  def valid_placement?(ship, placement_array)
+    placement_valid_length?(ship, placement_array) &&
```
Comment (jasonnoble) [inline lib/board.rb:56]: I've found it is difficult to troubleshoot why a placement is valid.

```
def valid_placement?(ship, coordinates)
    return false if ship.length != coordinates.length
    return false if coordinates.any?{|coordinate| !valid_coordinate?(coordinate)}
    return false if coordinates.any?{|coordinate| !cells[coordinate].empty? }
    return true if coordinates_in_same_row?(coordinates) && columns_consecutive?(coordinates)
    return true if coordinates_in_same_column?(coordinates) && rows_consecutive?(coordinates)
    return false
  end
```

Would you find this code easier to debug?  🤷

File: lib/board.rb [hunk 3]
```diff
@@ -6,9 +6,82 @@ class Board
...
+
+  def place(ship, placement_array)
+    placement_array.each do |coord|
+      @cells[coord].place_ship(ship)
+    end
+  end
+
+  def render(display_ship = false)
+    "  1 2 3 4 \n" +
+    "A " + @cells["A1"].render(display_ship) + " " + @cells["A2"].render(display_ship) + " " + @cells["A3"].render(display_ship) + " " + @cells["A4"].render(display_ship) + " \n" +
```
Comment (jasonnoble) [inline lib/board.rb:69]: You could DRY (Don't repeat yourself) up your code a bit here.  Maybe create a `render_row(row, display)` method?  `render_row('A', display)` would return all the cells that are in the A row, joined together by a space.

**Pull request merged**